File: packages/ui/db/entity/Trail.ts
import { cascadeOneToMany } from "@/db/cascadeOptions";
import BaseColumnSchemaPart from "@/db/entity/BaseColumnSchemaPart";
import { TrailLine } from "@/db/entity/TrailLine";
import { type ITrail } from "models";
import { EntitySchema } from "typeorm";

export type Trail = ITrail & {
  lines: TrailLine[];
};

const TrailSchema = new EntitySchema<Trail>({
  name: "trail",
  columns: {
    ...BaseColumnSchemaPart,
  },
  relations: {
    lines: {
      type: "one-to-many",
      inverseSide: "trail",
      target: "trail_line",
      ...cascadeOneToMany,
    },
    crag: {
      type: "one-to-one",
      target: "crag",
      inverseSide: "trail",
    },
  },
});

export default TrailSchema;

File: go-server/models/trail.go
package models

type Trail struct {
	ID    *int   `json:"id,omitzero"`
	Crag  *Crag  `json:"crag,omitzero"`
	Lines []Line `json:"lines,omitzero"`
}

func (t Trail) IsZero() bool {
	return t.ID == nil && t.Crag == nil && len(t.Lines) == 0
}
